### DISCUSSION TITLE
How to suppress moiré patterns during osgb files rendering

### DISCUSSION THREAD

**Discussion by wongshek** [openscenegraph/OpenSceneGraph#1298] [Q&A] [answered]
### Question
When I used osg to render an `osgb` file containing a large number of striped textures, I found that a very significant moiré phenomenon appeared when scaling, panning, and rotating. But when I render the `obj` file of the same scene with the same code, the moiré disappears. I guess some built-in texture filtering is done when rendering the `obj` file, so what should I do for the `osgb` file to suppress moiré?

![image](https://github.com/openscenegraph/OpenSceneGraph/assets/87893723/a2502b85-308f-41fb-be81-09a8aa03b334)

### Files(osgb+obj)
[MoErWen.zip](https://github.com/openscenegraph/OpenSceneGraph/files/14079779/MoErWen.zip)

### Code
```c++
#include <osgDB/ReadFile>
#include <osgViewer/Viewer>
#include <osgViewer/ViewerEventHandlers>
#include <osgGA/StateSetManipulator>
#include <osg/DisplaySettings>

int main()
{
    osg::DisplaySettings::instance()->setNumMultiSamples(4);

    osg::ref_ptr<osg::Node> n = osgDB::readNodeFile(osgPath);
    osg::ref_ptr<osg::Group> root = new osg::Group;
    osg::ref_ptr<osgViewer::Viewer> viewer = new osgViewer::Viewer;

    root->addChild(n);
    viewer->setSceneData(root);
    viewer->setUpViewInWindow(20, 20, 1400, 700);

    viewer->addEventHandler(new osgGA::StateSetManipulator(viewer->getCamera()->getOrCreateStateSet()));
    viewer->addEventHandler(new osgViewer::WindowSizeHandler);
    viewer->addEventHandler(new osgViewer::StatsHandler);
    viewer->realize();
   
    return viewer->run();
}
```

**Comment by robertosfield** [comment, marked as answer]
Convert both the .obj and .ossgb to .osgt then compare them both to see what the differences are with the various settings.  This should tell you what is going on. My guess the .obj loader is defaulting to mipmapping while the .osfb dataset has been set up without mimapping for some reason.

**Reply by wongshek** [reply]
Thanks for your quick reply. I converted both files to `osgt` for comparison, and found out that the `MIN_FILTER` field was inconsistent. I changed the `MIN_LILTER` field of the osgb file from `LINEAR` to `LINEAR_MIPMAP_LINEAR`, and the moiré pattern was greatly reduced!

In addition, I found that there are many other fields that are different between the two, such as `DataVariance`, `WRAP_S`. Forgive me for being a beginner, where should I go to learn the details of osg format?

Last question, the osgb format is still the most popular 3D results format in the field of photogrammetry. If I migrate the rendering engine from `osg` to `vsg`, do I need to change `osgb` to another format? Or is there a more recommended format for large-scale 3D data at the city level?

**Reply by AnyOldName3** [reply]
Data variance declares whether the data might change every frame (and therefore the OSG has to finish rendering everything that uses that object in a frame before preparing the next one, reducing/eliminating the opportunity for the main traversals and draw traversal to happen in parallel). Other than that, most of what you'll find in there is fairly self-explanatory (e.g. `osg::Group` is a group of nodes), or OpenGL fixed-function pipeline state that should be easily googleable.

**Reply by robertosfield** [reply]
The texture filters stored in the .osgb will have been set by the application that loaded them or from the original model, so the difference to the .obj is really just down to the history of that .osgb.  The default the OSG will normally opt for is to use mapping like you've found when loading the .obj.

As for .osgb and photgrammetry, this is not something I was ever aware of being an association until a couple of poster suggest this recently.  I can only guess that a popular application or library in the area of phtogrammetry used the OSG and just stored it's data as a .osgb, there is no intrinsic suitability of the .osgb for this, it's just the OSG's native format rather than some formal standard that one would normally use for exchanging data.

What photogrammetry application/library are you trying to work with?

If working with the VSG then you can use the [osg2vsg](https://github.com/vsg-dev/osg2vsg) format to load OSG scene graphs (and it's supported formats) then convert to a VSG scene graph, then do with what you want.  osg2vsg was written in the VSG's life so is pretty crude, it needs a rewrite to update it to use modern VSG capabilities.

Personally if I were to write a photogrammetry tool today I'd use a format like glTF for exchanging data, though this would required both reading and writing support of course.   The VSG has glTF support for loading rather than the use of the assimp library, but long term I'd like to have a read/write capability included in the VSG's vsgXchange lbrary.

**Reply by wongshek** [reply]
@AnyOldName3 
Thanks, I think I figured it out!

@robertosfield
One of the requirements of photogrammetry is the ability to smoothly navigate large-scale 3D models, which requires some 3D format that supports LOD.

In the past, ContextCapture was the mainstream software in the field of photogrammetry, and the beginning of the association between the osgb format and photogrammetry may be related to it. 

Now that more software has emerged, the 3dTiles format based on glTF is gradually becoming popular. How does vsg support 3dTiles?

**Reply by robertosfield** [reply]
For 3DTiles support you can use @timoore vsgCs library that integrates VulkanSceneGraph with cesium-ion:

https://github.com/timoore/vsgCs

**Reply by wongshek** [reply]
Thanks, I'll get into more details.
